application/src/ext-test/java/org/opentripplanner/ext/fares/impl/gtfs/CostedTransferInNetworkTest.java

```diff
@@ -0,0 +1,140 @@
+import org.opentripplanner.ext.fares.impl._support.FareTestConstants;
+import org.opentripplanner.ext.fares.model.FareLegRule;
+import org.opentripplanner.ext.fares.model.FareTransferRule;
+import org.opentripplanner.model.fare.FareOffer;
+import org.opentripplanner.model.fare.FareProduct;
+import org.opentripplanner.model.plan.PlanTestConstants;
+import org.opentripplanner.transit.model._data.TimetableRepositoryForTest;
+import org.opentripplanner.transit.model.basic.Money;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.network.GroupOfRoutes;
+import org.opentripplanner.transit.model.network.Route;
+
+class CostedTransferInNetworkTest implements PlanTestConstants, FareTestConstants {
+
+  private static final GroupOfRoutes NETWORK = groupOfRoutes("n1").build();
+  private static final Route ROUTE_1 = route("r1");
+  private static final Route ROUTE_2 = route("r2");
+  private static final Route ROUTE_3 = TimetableRepositoryForTest.route("r3").build();
+  private static final FeedScopedId LEG_GROUP = id("leg-group-a");
+  private static final FareProduct ADULT_PRODUCT = FareProduct.of(
+    id("adult"),
+    "adult",
+    Money.euros(10)
+  ).build();
+  private static final FareProduct YOUTH_PRODUCT = FareProduct.of(
+    id("youth"),
+    "youth",
+    Money.euros(5)
+  ).build();
+
+  private static final GtfsFaresV2Service SERVICE = new GtfsFaresV2Service(
+    List.of(
+      FareLegRule.of(id("r1"), ADULT_PRODUCT)
+        .withLegGroupId(LEG_GROUP)
+        .withNetworkId(NETWORK.getId())
+        .build(),
+      FareLegRule.of(id("r2"), YOUTH_PRODUCT)
+        .withLegGroupId(LEG_GROUP)
+        .withNetworkId(NETWORK.getId())
+        .build()
+    ),
+    List.of(
+      new FareTransferRule(
+        id("transfer"),
+        LEG_GROUP,
+        LEG_GROUP,
+        UNLIMITED_TRANSFERS,
+        null,
+        List.of(ONE_EUR_TRANSFER)
+      )
+    ),
+    Multimaps.forMap(Map.of())
+  );
+
+  @Test
+  void twoLegs() {
+    var i1 = newItinerary(A, 0).bus(ROUTE_1, 1, 0, 20, B).bus(ROUTE_2, 2, 21, 40, C).build();
+
+    var result = SERVICE.calculateFares(i1);
+
+    assertThat(result.itineraryProducts()).isEmpty();
+    assertThat(result.productsForLeg(i1.legs().getFirst())).containsExactly(
+      FareOffer.of(ADULT_PRODUCT),
+      FareOffer.of(YOUTH_PRODUCT)
+    );
+    assertThat(result.productsForLeg(i1.legs().getLast())).containsExactly(
+      FareOffer.of(ONE_EUR_TRANSFER, Set.of(ADULT_PRODUCT, YOUTH_PRODUCT)),
+      FareOffer.of(ADULT_PRODUCT),
+      FareOffer.of(YOUTH_PRODUCT)
+    );
+  }
+
+  @Test
+  void threeLegs() {
+    var i1 = newItinerary(A, 0)
+      .bus(ROUTE_1, 1, 0, 20, B)
+      .bus(ROUTE_2, 2, 21, 40, C)
+      .bus(ROUTE_2, 3, 41, 45, D)
+      .build();
+
+    var result = SERVICE.calculateFares(i1);
+
+    assertThat(result.itineraryProducts()).isEmpty();
+    assertThat(result.productsForLeg(i1.legs().getFirst())).containsExactly(
```

Comment:
Would it make sense to rename `.productsForLeg` => `.fareOffersForLeg`?

Reply:
It would